plugins/cxaudio/fu-cxaudio-device.c

```diff
@@ -0,0 +1,881 @@
+	if (operation == FU_CXAUDIO_OPERATION_WRITE &&
+	    mem_kind == FU_CXAUDIO_MEM_KIND_CPX_ROM) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "trying to write unwritable section %u",
+			     mem_kind);
+		return FALSE;
+	}
+
+	/* check memory address - should be within valid range */
+	switch (mem_kind) {
+	case FU_CXAUDIO_MEM_KIND_EEPROM:
+		size = 0x20000;
+		break;
+	default:
+		break;
+	}
+	if (addr > size) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "address out of range 0x%x < 0x%x",
+			     addr, size);
+		return FALSE;
+	}
+
+	/* send to hardware */
+	chunks = fu_chunk_array_new (buf, bufsz, addr, 0x0, payload_max);
+	for (guint i = 0; i < chunks->len; i++) {
+		FuChunk *chunk = g_ptr_array_index (chunks, i);
+		guint8 inbuf[FU_CXAUDIO_INPUT_REPORT_SIZE] = { 0 };
+		guint8 outbuf[FU_CXAUDIO_OUTPUT_REPORT_SIZE] = { 0 };
+
+		/* first byte is always report ID */
+		outbuf[0] = FU_CXAUDIO_MEM_WRITEID;
+
+		/* set memory address and payload length (if relevant) */
+		if (chunk->address >= 64 * 1024)
+			outbuf[1] |= 1 << 4;
+		outbuf[2] = chunk->data_sz;
+		fu_common_write_uint16 (outbuf + 3, chunk->address, G_BIG_ENDIAN);
+
+		/* set the write bit */
+		if (operation == FU_CXAUDIO_OPERATION_WRITE)
+			outbuf[1] |= 1 << 6;
+
+		/* set memtype */
+		if (mem_kind == FU_CXAUDIO_MEM_KIND_EEPROM)
+			outbuf[1] |= 1 << 5;
+
+		/* fill the report payload part */
+		if (operation == FU_CXAUDIO_OPERATION_WRITE) {
+			if (!fu_memcpy_safe (outbuf, sizeof(outbuf), idx_write, /* dst */
+					     chunk->data, chunk->data_sz, 0x0, /* src */
+					     chunk->data_sz, error))
+				return FALSE;
+		}
+		if (!fu_cxaudio_device_output_report (self, outbuf, sizeof(outbuf), error))
+			return FALSE;
+
+		/* issue additional write directive to read */
+		if (operation == FU_CXAUDIO_OPERATION_WRITE &&
+		    flags & FU_CXAUDIO_OPERATION_FLAG_VERIFY) {
+			outbuf[1] &= ~(1 << 6);
+			if (!fu_cxaudio_device_output_report (self, outbuf, sizeof(outbuf), error))
+				return FALSE;
+		}
+		if (operation == FU_CXAUDIO_OPERATION_READ ||
+		    flags & FU_CXAUDIO_OPERATION_FLAG_VERIFY) {
+			if (!fu_cxaudio_device_input_report (self,
+							     FU_CXAUDIO_MEM_READID,
+							     inbuf, sizeof(inbuf),
+							     error))
+				return FALSE;
+		}
+		if (operation == FU_CXAUDIO_OPERATION_WRITE &&
+		    flags & FU_CXAUDIO_OPERATION_FLAG_VERIFY) {
+			if (!fu_common_bytes_compare_raw (outbuf + idx_write, payload_max,
+							  inbuf + idx_read, payload_max,
+							  error)) {
+				g_prefix_error (error,
+						"failed to verify on packet %u @0x%x: ",
+						chunk->idx, chunk->address);
+				return FALSE;
+			}
+		}
+		if (operation == FU_CXAUDIO_OPERATION_READ) {
+			if (!fu_memcpy_safe ((guint8 *) chunk->data, chunk->data_sz, 0x0, /* dst */
+					     inbuf, sizeof(inbuf), idx_read, /* src */
+					     chunk->data_sz, error))
+				return FALSE;
+		}
```

Comment:
Actually I should scroll further, looks like you can probably optimize all 4 :)

Reply:
IE if it's read and verify do this, if it's read and not verify do that, and then if it's write do this other thing.